validate/validate.go

```diff
@@ -139,78 +148,83 @@ func (v *Validator) CheckRoot() (msgs []string) {
 		rootfsPath = filepath.Join(v.bundlePath, v.spec.Root.Path)
 		absRootPath, err = filepath.Abs(rootfsPath)
 		if err != nil {
-			msgs = append(msgs, fmt.Sprintf("unable to convert %q to an absolute path", rootfsPath))
+			errs = multierror.Append(errs, fmt.Errorf("unable to convert %q to an absolute path", rootfsPath))
 		}
 	}
 
 	if fi, err := os.Stat(rootfsPath); err != nil {
-		msgs = append(msgs, fmt.Sprintf("Cannot find the root path %q", rootfsPath))
+		errs = multierror.Append(errs,
+			NewError(PathExistence, fmt.Sprintf("Cannot find the root path %q", rootfsPath), rspec.Version))
 	} else if !fi.IsDir() {
-		msgs = append(msgs, fmt.Sprintf("The root path %q is not a directory.", rootfsPath))
+		errs = multierror.Append(errs,
+			NewError(PathExistence, fmt.Sprintf("The root path %q is not a directory.", rootfsPath), rspec.Version))
 	}
 
 	rootParent := filepath.Dir(absRootPath)
 	if absRootPath == string(filepath.Separator) || rootParent != absBundlePath {
-		msgs = append(msgs, fmt.Sprintf("root.path is %q, but it MUST be a child of %q", v.spec.Root.Path, absBundlePath))
+		errs = multierror.Append(errs,
+			NewError(ArtifactsInSingleDir, fmt.Sprintf("root.path is %q, but it MUST be a child of %q", v.spec.Root.Path, absBundlePath), rspec.Version))
 	}
 
 	if v.platform == "windows" {
 		if v.spec.Root.Readonly {
-			msgs = append(msgs, "root.readonly field MUST be omitted or false when target platform is windows")
+			errs = multierror.Append(errs,
+				NewError(ReadonlyOnWindows, "root.readonly field MUST be omitted or false when target platform is windows", rspec.Version))
 		}
 	}
 
 	return
 }
 
 // CheckSemVer checks v.spec.Version
-func (v *Validator) CheckSemVer() (msgs []string) {
+func (v *Validator) CheckSemVer() (errs error) {
 	logrus.Debugf("check semver")
 
 	version := v.spec.Version
 	_, err := semver.Parse(version)
 	if err != nil {
-		msgs = append(msgs, fmt.Sprintf("%q is not valid SemVer: %s", version, err.Error()))
+		errs = multierror.Append(errs,
+			NewError(SpecVersion, fmt.Sprintf("%q is not valid SemVer: %s", version, err.Error()), rspec.Version))
 	}
 	if version != rspec.Version {
-		msgs = append(msgs, fmt.Sprintf("internal error: validate currently only handles version %s, but the supplied configuration targets %s", rspec.Version, version))
+		errs = multierror.Append(errs, fmt.Errorf("internal error: validate currently only handles version %s, but the supplied configuration targets %s", rspec.Version, version))
```

Comment:
This is clearly an internal error because it is a non-RFC error.  I think we should drop “internal error: ” from the value.